### PR TITLE
New vis metrics

### LINKED ISSUE #1634: [FEAT] metrics for object detection + integrate into class performance
**Is your feature request related to a problem? Please describe.**
There are no metrics to measure performance of object detection model for a specific threshold. 

**Describe the solution you'd like**
Implement a set of metrics based on the logic described in https://docs.google.com/document/d/10cfF8HyQ_tau6D1k7Ip-i3aBedx_VP_B1VRDtCW3ddE/edit. As first step only implement TPR, FNR and FPR.
Add support in those metrics to the class performance check.

Add unit tests to make sure those metrics return the correct value
Add tests based on those metric to the vision class performance check

### PR DESCRIPTION
**Pull request by JKL98ISR**
closes #1634

### CODE REVIEW

**Comment by Nadav-Barak** [comment]
Would also add a demonstration of this feature to plot_class_performance description file

File: deepchecks/vision/checks/model_evaluation/class_performance.py [hunk 1]
```diff
@@ -54,6 +54,11 @@ class ClassPerformance(TrainTestCheck):
     class_list_to_show: List[int], default: None
         Specify the list of classes to show in the report. If specified, n_to_show, show_only and metric_to_show_by
         are ignored.
+    scorers_to_use : List[str], default: None
```
Comment (Nadav-Barak) [inline deepchecks/vision/checks/model_evaluation/class_performance.py:57]: Should also add arguments for the user to supply his selected IoU and Threshold if he wants to
(we would use the same IoU and threshold for all metrics in list)

File: deepchecks/vision/metrics_utils/metric_mixin.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    @abstractmethod
+    def get_labels_areas(self, labels) -> t.List[int]:
+        """Get labels object of single image and should return area for each label."""
+        pass
+
+
+class ObjectDetectionMetricMixin(MetricMixin):
+    """Metric util function mixin for object detection.
+
+    Expecting to receive the predictions in the following format: [x, y, w, h, confidence, label].
```
Comment (TheSolY) [inline deepchecks/vision/metrics_utils/metric_mixin.py:52]: If there's a name for the format (YOLO?) mention it

File: tests/vision/utils_tests/metrics_test.py [hunk 3]
```diff
@@ -12,7 +12,7 @@
 
 from deepchecks.vision import VisionData
 from deepchecks.vision.metrics_utils.metrics import calculate_metrics
-from deepchecks.vision.metrics_utils.object_detection_precision_recall import ObjectDetectionAveragePrecision
+from deepchecks.vision.metrics_utils.detection_precision_recall import ObjectDetectionAveragePrecision
```
Comment (TheSolY) [inline tests/vision/utils_tests/metrics_test.py:15]: Why did we rename?

**Review by TheSolY — CHANGES_REQUESTED** [review]
Functionally OK but some changes are required in order to improve legibility

File: deepchecks/vision/metrics_utils/detection_matched_positive_calc.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+        self._i = 0
+
+    @reinit__is_reduced
+    def update(self, output):
+        """Update metric with batch of samples."""
+        y_pred, y = output
+
+        for detected, ground_truth in zip(y_pred, y):
+            if isinstance(detected, torch.Tensor):
+                detected = detected.cpu()
```
[the remark this answers is not in the document]
Comment (JKL98ISR) [reply deepchecks/vision/metrics_utils/detection_matched_positive_calc.py:68]: it is for calculation running on cpu, need to add detach tho

File: deepchecks/vision/metrics_utils/detection_precision_recall.py [hunk 5]
```diff
@@ -335,27 +336,16 @@ def get_classes_scores_at(self, res: np.ndarray, iou: float = None, area: str =
...
+class ObjectDetectionAveragePrecision(AveragePrecisionRecall, ObjectDetectionMetricMixin):
+    """Calculate average precision and recall for object detection.
+
+    Parameters
+    ----------
+    max_dets: Union[List[int], Tuple[int]], default: [1, 10, 100]
+        Maximum number of detections per class.
+    area_range: tuple, default: (32**2, 96**2)
+        Slices for small/medium/large buckets.
+    return_option: int, default: 0
```
[the remark this answers is not in the document]
Comment (JKL98ISR) [reply deepchecks/vision/metrics_utils/detection_precision_recall.py:349]: it is not code from this pr, just rename the file

File: tests/vision/utils_tests/metrics_test.py (hunk 3 above)
Comment (JKL98ISR) [reply tests/vision/utils_tests/metrics_test.py:15]: because it's not specifically object detection

File: deepchecks/vision/metrics_utils/detection_precision_recall.py [hunk 6]
```diff
@@ -335,27 +336,16 @@ def get_classes_scores_at(self, res: np.ndarray, iou: float = None, area: str =
...
+
+class ObjectDetectionAveragePrecision(AveragePrecisionRecall, ObjectDetectionMetricMixin):
+    """Calculate average precision and recall for object detection.
+
+    Parameters
+    ----------
+    max_dets: Union[List[int], Tuple[int]], default: [1, 10, 100]
+        Maximum number of detections per class.
+    area_range: tuple, default: (32**2, 96**2)
+        Slices for small/medium/large buckets.
```
[the remark this answers is not in the document]
Comment (JKL98ISR) [reply deepchecks/vision/metrics_utils/detection_precision_recall.py:348]: it is the buckets for the area range

File: deepchecks/vision/checks/model_evaluation/class_performance.py (hunk 1 above)
Comment (JKL98ISR) [reply deepchecks/vision/checks/model_evaluation/class_performance.py:57]: should be a new issue for this feature

File: deepchecks/vision/metrics_utils/detection_precision_recall.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (JKL98ISR) [reply deepchecks/vision/metrics_utils/detection_precision_recall.py]: fixed it

File: deepchecks/vision/checks/model_evaluation/class_performance.py (hunk 1 above)
Comment (Nadav-Barak) [reply deepchecks/vision/checks/model_evaluation/class_performance.py:57]: I think it is insufficient to offer the user to give a metric without giving him the option to control its key elements (IoU, Threshold).
Decide how urgent is it @ItayGabbay @shir22

File: deepchecks/vision/metrics_utils/detection_tp_fp_fn_calc.py [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+import numpy as np
+import torch
+from ignite.metrics import Metric
+from ignite.metrics.metric import reinit__is_reduced, sync_all_reduce
+
+from deepchecks.vision.metrics_utils.evalution_functions import AVAILABLE_EVALUTING_FUNCTIONS
+from deepchecks.vision.metrics_utils.metric_mixin import MetricMixin, ObjectDetectionMetricMixin
+
+
+class TpFpFn(Metric, MetricMixin):
```
[the remark this answers is not in the document]
Comment (TheSolY) [reply deepchecks/vision/metrics_utils/detection_tp_fp_fn_calc.py:24]: @matanper maybe ConfusionMatrixCountsMixin because it's the counts and not the actual matrix. Though I think that TpFpFn is also fine.

File: deepchecks/vision/metrics_utils/evalution_functions.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+# Copyright (C) 2021-2022 Deepchecks (https://www.deepchecks.com)
+#
+# This file is part of Deepchecks.
+# Deepchecks is distributed under the terms of the GNU Affero General
+# Public License (version 3 or later).
+# You should have received a copy of the GNU Affero General Public License
+# along with Deepchecks.  If not, see <http://www.gnu.org/licenses/>.
+# ----------------------------------------------------------------------------
+#
+"""Module for evalution functions."""
```
Comment (TheSolY) [inline deepchecks/vision/metrics_utils/evalution_functions.py:11]: how are evaluation functions different from metrics? Maybe call it confusion_matrix_counts_metrics?

File: tests/vision/utils_tests/metrics_test.py (hunk 3 above)
Comment (TheSolY) [reply tests/vision/utils_tests/metrics_test.py:15]: what other types of detection are there?

File: deepchecks/vision/metrics_utils/detection_matched_positive_calc.py [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+                if g_idx in ground_truth_matched:
+                    continue
+
+                if ious[d_idx, g_idx] >= best_iou:
+                    best_iou = ious[d_idx, g_idx]
+                    best_match = g_idx
+            if best_match != -1:
+                detection_matches[d_idx] = best_match
+                ground_truth_matched[best_match] = d_idx
+        return detection_matches
```
[the remark this answers is not in the document]
Comment (JKL98ISR) [reply deepchecks/vision/metrics_utils/detection_matched_positive_calc.py]: idk what to call it

**Review by TheSolY — APPROVED** [review]
All comments fixed 3/7/22

**Review by matanper — APPROVED** [review]
